File: bench/cpp/ops/BenchPairwiseMatcher.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../CppBenchUtils.hpp"
#include "ops/generated/BenchPairwiseMatcherConfig.hpp"

#include <cvcuda/OpPairwiseMatcher.hpp>

#include <nvbench/nvbench.cuh>

template<typename ST>
inline void pairwisematcher(nvbench::state &state, nvbench::type_list<ST>)
try
{
    int3 shape = benchutils::GetShape<3, int3>(state.get_string("shape"));

    int matchesPerPoint = benchutils::GetIntParam<int>(state, "matchesPerPoint");

    bool crossCheck     = state.get_string("crossCheck") == "T";
    bool readNumSets    = state.get_string("readNumSets") == "T";
    bool writeDistances = state.get_string("writeDistances") == "T";

    NVCVNormType normType = benchutils::GetNormType(state.get_string("normType"));

    NVCVPairwiseMatcherType algoChoice;

    if (state.get_string("algoChoice") == "BRUTE_FORCE")
    {
        algoChoice = NVCV_BRUTE_FORCE;
    }
    else
    {
        throw std::invalid_argument("Unexpected algorithm choice = " + state.get_string("algoChoice"));
    }

    int maxMatches = shape.y * matchesPerPoint;

    cvcuda::PairwiseMatcher op(algoChoice);

    state.add_global_memory_reads((crossCheck ? 3 : 2) * shape.x * shape.y * shape.z * sizeof(ST));
    state.add_global_memory_writes(shape.x * (sizeof(int) + maxMatches * (2 * sizeof(int) + sizeof(float))));

    // clang-format off

    nvcv::Tensor set1({{shape.x, shape.y, shape.z}, "NMD"}, benchutils::GetDataType<ST>());
    nvcv::Tensor set2({{shape.x, shape.y, shape.z}, "NMD"}, benchutils::GetDataType<ST>());

    nvcv::Tensor matches({{shape.x, maxMatches, 2}, "NMD"}, nvcv::TYPE_S32);

    nvcv::Tensor numMatches({{shape.x}, "N"}, nvcv::TYPE_S32);

    nvcv::Tensor numSet1;
    nvcv::Tensor numSet2;
    nvcv::Tensor distances;

    if (readNumSets)
    {
        numSet1 = nvcv::Tensor({{shape.x}, "N"}, nvcv::TYPE_S32);
        numSet2 = nvcv::Tensor({{shape.x}, "N"}, nvcv::TYPE_S32);

        benchutils::FillTensor<int>(numSet1, [&shape](auto &){ return shape.y; });
        benchutils::FillTensor<int>(numSet2, [&shape](auto &){ return shape.y; });
    }
    if (writeDistances)
    {
        distances = nvcv::Tensor({{shape.x, maxMatches}, "NM"}, nvcv::TYPE_F32);
    }

    benchutils::FillTensor<ST>(set1, benchutils::LcgValues<ST>());
    benchutils::FillTensor<ST>(set2, benchutils::LcgValues<ST>());

    benchutils::warmup_and_exec(state, BENCH_PAIRWISEMATCHER_WARMUP_ITERATIONS,
        [&op, &set1, &set2, &numSet1, &numSet2, &matches, &numMatches, &distances, &crossCheck,
         &matchesPerPoint, &normType](cudaStream_t s) {
            op(s, set1, set2, numSet1, numSet2, matches, numMatches, distances, crossCheck,
               matchesPerPoint, normType);
        });
}
CVCUDA_BENCH_SKIP_ERRORS(state)

// clang-format on

// Use auto-generated type list from bench_params.json

NVBENCH_BENCH_TYPES(pairwisematcher, NVBENCH_TYPE_AXES(BENCH_PAIRWISEMATCHER_TYPES))
BENCH_PAIRWISEMATCHER_AXES;

File: bench/python/ops/bench_pairwisematcher.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import cupy as cp  # noqa: E402
import cvcuda  # noqa: E402
from python_bench_utils import (  # noqa: E402
    get_dtype,
    get_dtype_size,
    create_tensor,
    create_stream_cache,
    run_benchmark,
)


def pairwisematcher(state):
    """PairwiseMatcher operator benchmark matching C++ BenchPairwiseMatcher.cpp"""

    shape_str = state.get_string("shape")
    matches_per_point = state.get_int64("matchesPerPoint")
    cross_check = state.get_string("crossCheck") == "T"
    read_num_sets = state.get_string("readNumSets") == "T"
    write_distances = state.get_string("writeDistances") == "T"
    norm_type_str = state.get_string("normType")
    dtype = get_dtype(state.get_string("InOutDataType"))
    device_id = state.get_device()

    parts = shape_str.split("x")
    N = int(parts[0])
    M = int(parts[1])
    D = int(parts[2])
    shape = (N, M, D)

    norm_map = {
        "HAMMING": cvcuda.Norm.HAMMING,
        "L1": cvcuda.Norm.L1,
        "L2": cvcuda.Norm.L2,
    }
    norm_type = norm_map.get(norm_type_str, cvcuda.Norm.L2)

    dtype_size = get_dtype_size(dtype)
    max_matches = M * matches_per_point
    state.add_global_memory_reads((3 if cross_check else 2) * N * M * D * dtype_size)
    state.add_global_memory_writes(N * (4 + max_matches * (2 * 4 + 4)))

    set1 = create_tensor(shape, dtype, device_id, layout="NMD", fill_mode="lcg")
    set2 = create_tensor(shape, dtype, device_id, layout="NMD", fill_mode="lcg")

    num_set1 = None
    num_set2 = None
    if read_num_sets:
        num_set1_data = cp.full((N,), M, dtype=cp.int32)
        num_set1 = cvcuda.as_tensor(num_set1_data, "N")
        num_set2_data = cp.full((N,), M, dtype=cp.int32)
        num_set2 = cvcuda.as_tensor(num_set2_data, "N")

    get_stream = create_stream_cache()

    def run(launch):
        stream = get_stream(launch)
        cvcuda.match(
            set1,
            set2,
            num_set1=num_set1,
            num_set2=num_set2,
            num_matches=True,
            distances=write_distances,
            cross_check=cross_check,
            matches_per_point=matches_per_point,
            norm_type=norm_type,
            algo_choice=cvcuda.Matcher.BRUTE_FORCE,
            stream=stream,
        )

    return run


if __name__ == "__main__":
    run_benchmark("pairwisematcher", pairwisematcher)
